Tests/RevenueCatUITests/Templates/ExternalPurchaseAndRestoreTests.swift

```diff
@@ -0,0 +1,393 @@
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ info in
+            callbackOrder.append("onRestoreCompleted")
+            customerInfo = info
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.restorePurchases()
+            completed = true
+        }
+
+        expect(completed).toEventually(beTrue())
+        expect(customerInfo).toEventually(be(TestData.customerInfo))
+        expect(callbackOrder).toEventually(equal(["onRestoreStarted", "performRestore", "onRestoreCompleted"]))
+    }
+
+    func testHandleExternalRestoreFailure() throws {
+        var callbackOrder = [String]()
+        var passedError: Error?
+
+        let purchasHandler = Self.externalPurchaseHandler { _ in
+            return (userCancelled: true, error: nil)
+        } performRestore: {
+            callbackOrder.append("performRestore")
+            return (success: false, error: TestError.error)
+        }
+
+        try PaywallView(
+            offering: Self.offering.withLocalImages,
+            customerInfo: TestData.customerInfo,
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { _ in
+            callbackOrder.append("onPurchaseFailure")
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ _ in
+            callbackOrder.append("onRestoreCompleted")
+        })
+        .onRestoreFailure({ error in
+            callbackOrder.append("onRestoreFailure")
+            passedError = error
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.restorePurchases()
+        }
+
+        expect(callbackOrder).toEventually(equal(["onRestoreStarted", "performRestore", "onRestoreFailure"]))
+        expect(passedError).toEventually(matchError(TestError.error))
+    }
+
+    func testHandleExternalPurchaseWithoutPurchaseHandler() throws {
+        var errorThrown = false
+
+        let purchasHandler = Self.externalPurchaseHandler()
+
+        let config = PaywallViewConfiguration(purchaseHandler: purchasHandler)
+
+        try PaywallView(configuration: config).addToHierarchy()
+
+        Task {
+            do {
+                _ = try await purchasHandler.purchase(package: Self.package)
+            } catch {
+                errorThrown = true
+            }
+        }
+
+        expect(errorThrown).toEventually(beTrue())
```

Comment:
This can be simplified with just async / await, but there are also some concurrency goodies in Nimble now that we could explore

https://github.com/Quick/Nimble/blob/6ba9e680a1962a86510e9ea2c5c79fd43c5eeff4/Sources/Nimble/Nimble.docc/Guides/Concurrency.md

Reply:
We're using Nimble 10, a lot of this stuff came in in Nimble 11, current version is Nimble 13. Definitely time for an update, but maybe as part of a separate PR?

Reply:
I started looking at this, and I don't think there are any complicated changes but there are a lot of them (they've described their own time interval type, NimbleTimeInterval), so I think this would be better done as a separate PR.